optuna_integration/botorch/botorch.py

```diff
@@ -662,6 +674,158 @@ def qparego_candidates_func(
+
+    train_x = normalize(train_x, bounds=bounds)
+    if pending_x is not None:
+        pending_x = normalize(pending_x, bounds=bounds)
+
+    model = SingleTaskGP(train_x, train_y, outcome_transform=Standardize(m=train_y.size(-1)))
+    mll = ExactMarginalLogLikelihood(model.likelihood, model)
+    fit_gpytorch_mll(mll)
+
+    acqf = qKnowledgeGradient(
+        model=model,
+        num_fantasies=256,
+        objective=objective,
+        X_pending=pending_x,
+    )
+
+    standard_bounds = torch.zeros_like(bounds)
+    standard_bounds[1] = 1
+
+    candidates, _ = optimize_acqf(
+        acq_function=acqf,
+        bounds=standard_bounds,
+        q=1,
+        num_restarts=10,
+        raw_samples=512,
+        options={"batch_limit": 5, "maxiter": 200},
+        sequential=True,
+    )
+
+    candidates = unnormalize(candidates.detach(), bounds=bounds)
+
+    return candidates
+
+
+@experimental_func("3.7.0")
+def qhvkg_candidates_func(
+    train_x: "torch.Tensor",
+    train_obj: "torch.Tensor",
+    train_con: Optional["torch.Tensor"],
+    bounds: "torch.Tensor",
+    pending_x: Optional["torch.Tensor"],
+) -> "torch.Tensor":
+    """Quasi MC-based batch Hypervolume Knowledge Gradient (qHVKG).
+
+    According to Botorch/Ax documentation,
+    this function may perform better than qEHVI (`qehvi_candidates_func`).
+    (cf. https://botorch.org/tutorials/decoupled_mobo )
+
+    .. seealso::
+        :func:`~optuna_integration.botorch.qei_candidates_func` for argument and return value
+        descriptions.
+    """
+
+    # We need botorch >=0.9.5 for qHypervolumeKnowledgeGradient.
+    if not _imports_qhvkg.is_successful():
+        raise ImportError(
+            "qhvkg_candidates_func requires botorch >=0.9.5. "
+            "Please upgrade botorch or use qehvi_candidates_func as candidates_func instead."
+        )
+
+    if train_con is not None:
+        train_y = torch.cat([train_obj, train_con], dim=-1)
+    else:
+        train_y = train_obj
+
+    train_x = normalize(train_x, bounds=bounds)
+    if pending_x is not None:
+        pending_x = normalize(pending_x, bounds=bounds)
+
+    models = [
+        SingleTaskGP(train_x, train_y[..., [i]], outcome_transform=Standardize(m=1))
+        for i in range(train_y.shape[-1])
+    ]
+    model = ModelListGP(*models)
+    mll = SumMarginalLogLikelihood(model.likelihood, model)
+    fit_gpytorch_mll(mll)
+
+    n_constraints = train_con.size(1) if train_con is not None else 0
+    objective = FeasibilityWeightedMCMultiOutputObjective(
+        model,
+        X_baseline=train_x,
+        constraint_idcs=[-n_constraints + i for i in range(n_constraints)],
+    )
+
+    ref_point = train_obj.min(dim=0).values - 1e-8
+
+    acqf = qHypervolumeKnowledgeGradient(
+        model=model,
+        ref_point=ref_point,
+        num_fantasies=16,
+        X_pending=pending_x,
+        objective=objective,
+        sampler=ListSampler(
```

Comment:
We set the sampler explicitly to support `botorch`==0.9.5  (replicates the changes from pytorch/botorch#2160).